### ISSUE TITLE
Is glm-4.7-flash MTP training in RL stage supported for now?

### ISSUE THREAD

**Issue by ifififa** [THUDM/slime#1534]
Hi dear authors, can we use `--mtp-num-layers 1 --enable-mtp-training` to train glm-4.7-flash in RL stage?

**Comment by zhuzilin** [comment]
It is supported.

**Comment by ifififa** [comment]
> It is supported.

But I still met some trouble when conduct mtp training of glm-4.7-flash. 
1. refer to https://github.com/THUDM/slime/pull/1460, glm4_moe_lite is registered as deepseek_v3, 
```python
from mbridge.core import register_model
from mbridge.models import DeepseekV3Bridge


@register_model("glm4_moe_lite")
class GLM4MoELiteBridge(DeepseekV3Bridge):
    pass
```
when set `--mtp-num-layers 1` during hf to torch_dist conversion, the assertion error is triggered at https://github.com/ISEEKYAN/mbridge/blob/78b4efd5bb82df64fa3af542ecd80e7f1144c4d9/mbridge/models/deepseek_v3.py#L361

2. I modify https://github.com/ISEEKYAN/mbridge/blob/main/mbridge/models/deepseek_v3.py by changing `model.layers.61` to  `model.layers.47` but it seems do not work. I met grad norm `NaN` and mtp loss `NaN` during training.

@huang3eng  can you give some ideas about how to add mtp support for glm-4.7-flash? I notice that you have ran mtp training of Qwen3-Next successfully (https://github.com/THUDM/slime/pull/1503). Good Job!

**Comment by ifififa** [comment]
Update, train loss seems normal when set cp=1.

**Comment by bcol23** [comment]
Experiencing the same issue with Qween3.5-35B-A3B. 

When enabling MTP training, both `grad_norm` and `mtp_loss` become `NaN`. Here is the training log:

```text
model.py:660 - step 1: {'train/loss': 1.087484359741211, 'train/grad_norm': nan, 'train/mtp_loss': nan, 'train/lr-pg_0': 0.0, 'train/lr-pg_1': 0.0, 'train/lr-pg_2': 0.0, 'train/step': 1}
```

**Comment by huang3eng** [comment]
@bcol23  For `Qwen3.5-35B-A3B`, I did **not** hit `NaN` on `grad_norm` or `mtp_loss` with the setup below. 

Maybe the key point missing, please make sure `--mtp-num-layers 1` is added to `/root/slime/scripts/models/qwen3.5-35B-A3B.sh` **before** doing the `hf -> torch_dist` conversion. If the checkpoint was already converted without this flag, please reconvert it.

**Environment**
1. Docker image: `slimerl/slime:v0.2.3`
2. Also update to `transformers==5.2.0`

**HF -> torch_dist conversion**
```bash
cd /root/slime

source scripts/models/qwen3.5-35B-A3B.sh

PYTHONPATH=/root/Megatron-LM/ torchrun --nproc-per-node 8 \
   tools/convert_hf_to_torch_dist.py \
   ${MODEL_ARGS[@]} \
   --hf-checkpoint /root/Qwen3.5-35B-A3B \
   --save /root/Qwen3.5-35B-A3B_torch_dist
```

**Training script**
```bash
#!/bin/bash

# for rerun the task
pkill -9 sglang
sleep 3
ray stop --force
pkill -9 ray
pkill -9 python
sleep 3
pkill -9 ray
pkill -9 python

set -ex

# will prevent ray from buffering stdout/stderr
export PYTHONBUFFERED=16

# unset proxy to avoid issues
unset http_proxy https_proxy HTTP_PROXY HTTPS_PROXY

NVLINK_COUNT=$(nvidia-smi topo -m 2>/dev/null | grep -o 'NV[0-9][0-9]*' | wc -l)
if [ "$NVLINK_COUNT" -gt 0 ]; then
    HAS_NVLINK=1
else
    HAS_NVLINK=0
fi
echo "HAS_NVLINK: $HAS_NVLINK (detected $NVLINK_COUNT NVLink references)"

SCRIPT_DIR="$(cd -- "$(dirname -- "${BASH_SOURCE[0]}")" &>/dev/null && pwd)"
source "${SCRIPT_DIR}/models/qwen3.5-35B-A3B.sh"

CKPT_ARGS=(
   --hf-checkpoint /root/Qwen3.5-35B-A3B
   --ref-load /root/Qwen3.5-35B-A3B_torch_dist
   --load /root/Qwen3.5-35B-A3B_slime/
   --save /root/Qwen3.5-35B-A3B_slime/
   --save-interval 20
   --no-save-optim
   --no-load-optim
)

ROLLOUT_ARGS=(
   --prompt-data /root/dapo-math-17k/dapo-math-17k.jsonl
   --input-key prompt
   --label-key label
   --apply-chat-template
   --rollout-shuffle
   --rm-type math
   --num-rollout 3000
   --rollout-batch-size 16
   --n-samples-per-prompt 8
   --rollout-max-response-len 8192
   --rollout-temperature 1

   --global-batch-size 128
   --balance-data
)

EVAL_ARGS=(
   --eval-interval 50
   --eval-prompt-data aime /root/aime-2024/aime-2024.jsonl
   --n-samples-per-eval-prompt 16
   --eval-max-response-len 16384
   --eval-top-p 1
   --skip-eval-before-train
)

PERF_ARGS=(
   --tensor-model-parallel-size 2
   --sequence-parallel
   --pipeline-model-parallel-size 1
   --context-parallel-size 4
   --expert-model-parallel-size 8
   --expert-tensor-parallel-size 1

   --recompute-granularity full
   --recompute-method uniform
   --recompute-num-layers 1

   --use-dynamic-batch-size
   --max-tokens-per-gpu 4096
)

GRPO_ARGS=(
   --advantage-estimator grpo
   --kl-loss-coef 0.00
   --kl-loss-type low_var_kl
   --entropy-coef 0.00
   --eps-clip 0.2
   --eps-clip-high 0.28
)

OPTIMIZER_ARGS=(
   --optimizer adam
   --lr 1e-6
   --lr-decay-style constant
   --weight-decay 0.1
   --adam-beta1 0.9
   --adam-beta2 0.98

   --optimizer-cpu-offload
   --overlap-cpu-optimizer-d2h-h2d
   --use-precision-aware-optimizer
)

WANDB_ARGS=(
)

SGLANG_ARGS=(
   --rollout-num-gpus-per-engine 8
   --sglang-mem-fraction-static 0.7
   --sglang-ep-size 8

   --sglang-cuda-graph-bs 1 2 4 8 $(seq 16 8 128)

   --sglang-speculative-algorithm EAGLE
   --sglang-speculative-num-steps 2
   --sglang-speculative-eagle-topk 1
   --sglang-speculative-num-draft-tokens 3

   --sglang-max-running-requests 128
)

MISC_ARGS=(
   --attention-dropout 0.0
   --hidden-dropout 0.0
   --accumulate-allreduce-grads-in-fp32
   --attention-softmax-in-fp32
   --attention-backend flash

   --moe-token-dispatcher-type flex
)

SPEC_ARGS=(
   --mtp-num-layers 1
   --enable-mtp-training
   --mtp-loss-scaling-factor 0.1
)

export MASTER_ADDR=${MASTER_ADDR:-"127.0.0.1"}
export no_proxy="127.0.0.1,${MASTER_ADDR}"
ray start --head --node-ip-address ${MASTER_ADDR} --num-gpus 8 --disable-usage-stats --dashboard-host=0.0.0.0 --dashboard-port=8265

RUNTIME_ENV_JSON="{
  \"env_vars\": {
    \"PYTHONPATH\": \"/root/Megatron-LM/\",
    \"CUDA_DEVICE_MAX_CONNECTIONS\": \"1\",
    \"NCCL_NVLS_ENABLE\": \"${HAS_NVLINK}\",
    \"no_proxy\": \"${no_proxy}\"
  }
}"

ray job submit --address="http://127.0.0.1:8265" \
   --runtime-env-json="${RUNTIME_ENV_JSON}" \
   -- python3 train.py \
   --actor-num-nodes 1 \
   --actor-num-gpus-per-node 8 \
   --colocate \
   ${MODEL_ARGS[@]} \
   ${CKPT_ARGS[@]} \
   ${ROLLOUT_ARGS[@]} \
   ${OPTIMIZER_ARGS[@]} \
   ${GRPO_ARGS[@]} \
   ${WANDB_ARGS[@]} \
   ${PERF_ARGS[@]} \
   ${EVAL_ARGS[@]} \
   ${SGLANG_ARGS[@]} \
   ${MISC_ARGS[@]} \
   ${SPEC_ARGS[@]} \
   2>&1 | tee qwen_35_debug.log
```

With the above setup, the metrics look normal and I do not see `NaN`:
```text
[2026-03-27 10:10:32] model.py:660 - step 0: {'train/loss': -2.3283064365386963e-10, 'train/pg_loss': -2.3283064365386963e-10, 'train/entropy_loss': 0.3812420666217804, 'train/pg_clipfrac': 0.0, 'train/ppo_kl': 0.0, 'train/train_rollout_logprob_abs_diff': 0.014618218876421452, 'train/grad_norm': 1.8856140440868185, 'train/mtp_loss': 0.46907463669776917, 'train/lr-pg_0': 1e-06, 'train/lr-pg_1': 1e-06, 'train/lr-pg_2': 1e-06, 'train/step': 0}
```
